### PR TITLE
Add support for primus-rooms

### PR DESCRIPTION
**Pull request by fadeenk**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -76,6 +76,15 @@ assumptions to the data that is sent to the server:
   spark id. If no `sparks` property is supplied we assume that the given message
   needs to be broadcasted to every single connection on this server.
 
+If `primus-rooms` is installed, you can use extra properties
+
+- The `rooms` Property can be an array or a string of room names. Message will be
```
Comment (lpinca) [inline README.md]: Why is property capitalized?

File: README.md [hunk 2]
```diff
@@ -76,6 +76,15 @@ assumptions to the data that is sent to the server:
...
+ broadcasted to rooms property, if no rooms are provided will broadcast the message
```
Comment (lpinca) [inline README.md]: Please remove "property", "broadcasted to all rooms, if ...".

File: omega.js [hunk 3]
```diff
@@ -116,16 +116,22 @@ function parse(primus, raw, res) {
...
   //         the data to.
   // Spark:  The data.sparks is the id of one single individual spark which
   //         should receive the data.
   // All:    Broadcast the message to every single connected spark if no
   //         `data.sparks` has been provided.
   //
-  if (Array.isArray(data.sparks)) {
+  if (primus.rooms && Array.isArray(data.rooms) || typeof data.rooms === 'string') {
+      primus.room(data.rooms).except(data.except).write(data.msg);
+      called = data.rooms;
```
Comment (lpinca) [inline omega.js]: This is the only problem I see in this PR as in this case `called` is not the number of sparks the message has been sent to but a string representing the rooms.

**Comment by lpinca** [comment]
This also needs tests. Thanks!

File: omega.js (hunk 3 above)
Comment (fadeenk) [inline omega.js]: Will updated so the return matches the current format. Will probably be done over the weekend.

**Comment by fadeenk** [comment]
I meant to make an issue to see if adding primus-rooms support aligns with the goals of this plugin. I will be polishing it up and adding tests asap

**Comment by fadeenk** [comment]
@lpinca there seems to be some issue with ioredis module for node v4.7.0, currently I am adding more test cases to bring the coverage back up

**Comment by lpinca** [comment]
@fadeenk thank you for the effort you put here.

Code mostly looks good to me but I'm bit conflicted as the change is pretty big (I thought it was simpler) and adds a significant cost of maintainability.

What is in your opinion the advantage of supporting this in `omega-supreme` instead of resolving the spark ids in the application and use `omega-supreme` as is?

```js
primus.room(rooms).clients((err, sparks) => {
  //
  // Filter out duplicated and excluded sparks. Use metroplex if needed.
  //
  primus.forward(server, msg, ids, fn);
});
```

This is basically moving all the complexity in `omega-supreme` which is nice for the user but not for the module.

cc: @3rd-Eden

**Comment by fadeenk** [comment]
I agree, I thought it was way simpler as well. As for maintainability I dont mind supporting these modules that I am contributing to, since not only I am using them personally but also using them at work.

The reasoning is to add support for broadcasting to specific rooms via http requests without having to add another module to the projects for the users.

File: omega.js [hunk 4]
```diff
@@ -104,28 +104,90 @@ function parse(primus, raw, res) {
...
   //
-  if (Array.isArray(data.sparks)) {
+  if (callingRooms) {
+    // initialize except array
+    data.except = data.except || [];
+    if (typeof data.except === 'string') {
+      data.except = [data.except];
+    }
+    //get the sparks in the rooms
+    primus.room(data.rooms).clients(function(err, sparks) {
```
Comment (lpinca) [inline omega.js]: Error is not handled here.

File: package.json [hunk 5]
```diff
@@ -5,6 +5,7 @@
   "main": "index.js",
   "scripts": {
     "coverage": "istanbul cover _mocha",
+    "coverage-win": "istanbul cover node_modules/mocha/bin/_mocha -- -R spec",
```
Comment (lpinca) [inline package.json]: Why is this needed?

File: test.js [hunk 6]
```diff
@@ -424,4 +428,258 @@ describe('omega supreme', function () {
...
+          });
+        }, function (err, ids) {
+          if (err) return done(err);
+          request({
+            url: url(server.url, '/primus/omega/supreme'),
+            auth: {user: 'omega', pass: 'supreme'},
+            method: 'PUT',
+            json: {msg: 'foo', rooms: rooms[0]}
+          }, function (err, res, body) {
+            if (err) return next(err);
```
Comment (lpinca) [inline test.js]: I think this should be `done` not `next`.

File: package.json (hunk 5 above)
Comment (fadeenk) [inline package.json]: To generate coverage reports on windows.... I develop on a windows machine

**Comment by lpinca** [comment]
I've added few comments, feel free to ignore them for now until we decide what to do with this.

File: package.json (hunk 5 above)
Comment (lpinca) [inline package.json]: Yes, I reformulate the question, why doesn't `npm run coverage` work on Windows?

Comment (fadeenk) [inline package.json]: Its an issue with how windows works with npm packages, it creates a batch script that does not work nicely with istanbul because it assumes it is loading a js file for more info you can take a look at the original issue
https://github.com/gotwarlost/istanbul/issues/90

Comment (lpinca) [inline package.json]: Thanks, does it work if the script is written like this?

```
istanbul cover node_modules/mocha/bin/_mocha
```

If so I would just change it like that.

Comment (fadeenk) [inline package.json]: fixed :)

Comment (lpinca) [inline package.json]: I mean to change the `coverage` script so we don't have one specifically for Windows and it works on all platforms :)

Comment (fadeenk) [inline package.json]: oh i see, I misread that. 

I believe it should work, I dont have access to my linux box right now so cant confirm but can change it and see how the build goes

**Comment by fadeenk** [comment]
I was thinking maybe we could just add support for custom parsers which will allow us to have plugin and play parsers for the different setups that the users may have. That will keep this module nice and clean, while allowing the user full control on how to handling parsing. 

That should work unless u have any objections, I will just have to do bunch of more work on this which could take a while

**Comment by lpinca** [comment]
@fadeenk It may be even simpler, for example the user could replace the `omega-supreme` middleware with a custom one.

```js
const index = primus.indexOfLayer('authorization');

primus.remove('omega-supreme');
primus.use('omega-supreme', customMiddleware, options, index);
```

**Comment by fadeenk** [comment]
That would be great will see what I can do, most likely wont be able to work on it much till next weekend though

**Comment by lpinca** [comment]
@fadeenk no problem. The added benefit is that the middleware could also be published as a separate module which is nice.
